examples/interpreter/physl.cpp

```diff
@@ -21,36 +22,97 @@
 namespace fs = boost::filesystem;
 namespace po = boost::program_options;
 
-std::string read_user_code(std::string const& arg)
+std::string read_user_code(std::string const& path)
 {
-    std::ifstream code_stream(arg);
+    std::ifstream code_stream(path);
     if (!code_stream.good())
     {
         HPX_THROW_EXCEPTION(hpx::filesystem_error,
             "read_user_code",
-            "Failed to open the specified file: " + arg);
+            "Failed to open the specified file: " + path);
     }
 
-    std::string code;
-    // Find out how much memory we need to allocate
-    code_stream.seekg(0, std::ios::end);
-    // Allocate all the needed memory upfront
-    code.reserve(code_stream.tellg());
-    // Back to the beginning of the file
-    code_stream.seekg(0, std::ios::beg);
-
     // Read the file
-    code.assign(std::istreambuf_iterator<char>(code_stream),
-        std::istreambuf_iterator<char>());
+    std::ostringstream str_stream;
+    if (!(str_stream << code_stream.rdbuf()))
+    {
+        HPX_THROW_EXCEPTION(hpx::filesystem_error,
+            "read_user_code",
+            "Failed to read code from the specified file: " + path);
+    }
 
-    return code;
+    return str_stream.str();
+}
+
+void dump_ast(std::vector<phylanx::ast::expression> ast, std::string path)
+{
+    std::ofstream ast_stream(path, std::ios::binary);
+    if (!ast_stream.good())
+    {
+        HPX_THROW_EXCEPTION(hpx::filesystem_error,
+            "dump_ast",
+            "Failed to open the specified file: " + path);
+    }
+
+    // Serialize the AST into a byte array
+    std::vector<char> bytes = phylanx::util::serialize(ast);
+
+    // char is 1 byte
+    ast_stream.write(bytes.data(), bytes.size());
+}
+
+std::vector<phylanx::ast::expression> load_ast(std::string path)
+{
+    std::ifstream str_stream(path, std::ios::binary);
+    auto const start_pos = str_stream.tellg();
+    if (std::streamsize(-1) == start_pos)
+    {
+        HPX_THROW_EXCEPTION(hpx::filesystem_error,
+            "load_ast",
+            "Failed to open the specified file: " + path);
+    }
+
+    if (!str_stream.ignore((std::numeric_limits<std::streamsize>::max)()))
```

Comment:
This will stop loading if `-1` (`EOF`) is found. What is the rationale of using `.ignore()`?